### ISSUE TITLE
How to render a mesh with camera intrinsics matrix

### ISSUE THREAD

**Issue by karta2155802** [facebookresearch/pytorch3d#1427]
Hi, thanks for the useful tool.
I encounter some problems with rendering an object with camera matrix.

I have a object mesh with its 3d coordinates from a dataset with camera intrinsic matrix. I think this dataset sets the camera at world's origin. It didn't provide camera's R,T.
I expect the rendered object should have the same pose and position with the original photo. Is it possible to achieve?
![未命名](https://user-images.githubusercontent.com/41221410/213616579-4009a06d-f2e2-4b87-b4ea-5ef661de1f09.png)

Here are some of the object vertices position:
[[-0.0227,  0.2144, -0.7908],
  [-0.0179,  0.2163, -0.7865],
  [-0.0400,  0.2093, -0.7738],
  ...,
  [-0.0952,  0.0324, -0.6633],
  [-0.0896,  0.0411, -0.6660],
  [-0.0864,  0.0482, -0.6599]]
camera intrinsic matrix:
[[617.287,   0.        , 314.564],
  [  0.       , 617.061, 236.353],
  [  0.       ,   0.        , 1.   ]]

I'm trying to render the object with the code below, but I end up getting an empty screen. I don't know how to set the camera correctly.
```
mesh = Meshes(
    verts=[verts.to(device)],
    faces=[faces.to(device)],
    textures=textures
)
focal = torch.tensor([[617.287, 617.061]], device=device)
princpt = torch.tensor([[314.564, 236.353]], device=device)
cameras = PerspectiveCameras(device=device, focal_length=focal, principal_point=princpt)
raster_settings = RasterizationSettings(
    image_size=(480,640),
    blur_radius=0.0,
    faces_per_pixel=1, 
)
lights = PointLights(device=device, location=[[0.0, 0.0, -3.0]])

renderer = MeshRenderer(
    rasterizer=MeshRasterizer(
        cameras=cameras, 
        raster_settings=raster_settings
    ),
    shader=HardPhongShader(
        device=device, 
        cameras=cameras,
        lights=lights
    )
)

images = renderer(mesh).cpu().numpy()
plt.imshow(images.squeeze()[:,:,:3])
plt.show()
```
If I switch to another type of camera, I can get the object, but not the view I want.
```
R, T = look_at_view_transform(0.5, 0, 0)
cameras = FoVPerspectiveCameras(device=device, R=R, T=T)
```
![342](https://user-images.githubusercontent.com/41221410/213616721-0938a03f-c55e-44d7-95e0-f9d2e46d1692.png)

**Comment by karta2155802** [comment]
After few days trying, I found out that I made a few terrible mistakes.
1. The vertices are in openGL coordinate, which is different from pytorch3D.
2. The focal length and principal point are in screen space, which should set parameter in PerspectiveCamera **in_ndc** as False, and give the image size as input.

After modifying my code, I can get the correct result, so now I close the issue.

**Comment by enjoybo** [comment]
@karta2155802 hi , could you please share the code of rendering the same results using opengl and pytorch3d ? I feel confused to reproduce the same pytorch3d rendering result using opengl. Thx.

**Comment by karta2155802** [comment]
I finally use this function for rendering object mask. 
"B" indicates batch size.

```
def render_mask(verts, verts_rgb, faces, focal, princpt, img_size):
    '''
    patam verts: list of vertex tensors in shape [N, 3]
    param verts_rgb: list of tensors, with same shape as verts
    param faces: list of face tensors in shape [num_faces, 3]
    param focal: [B, 2]
    param printcpt: [B, 2]
    param img_size: (h,w)
    return [B, 1, 240, 320] alpha channel of rendered images
    '''
    B = focal.size(0)
    device = focal.device
    textures=TexturesVertex(verts_features=verts_rgb)
    mesh = Meshes(
        verts=verts,
        faces=faces,
        textures=textures
    )

    cam_img_size = [img_size for i in range(B)]
    cameras = PerspectiveCameras(device=device, focal_length=focal, principal_point=princpt, in_ndc=False, image_size=cam_img_size)
    raster_settings = RasterizationSettings(image_size=img_size, blur_radius=0.0, faces_per_pixel=1)
    renderer = MeshRenderer(
        rasterizer=MeshRasterizer(cameras=cameras, raster_settings=raster_settings), 
        shader=SoftPhongShader(device=device, cameras=cameras)
    )

    images = renderer(mesh) # B x 480 x 640 x 4
    masks = (images[...,3]).unsqueeze(1) # B x 1 x 480 x 640
    masks = F.interpolate(masks, scale_factor=(0.5)) # downsample to B x 1 x 240 x 320

    return masks
```

If you cannot see the object, maybe you can try to change the vertices coordinate by multiplying [1, -1, -1].

**Comment by enjoybo** [comment]
这是来自QQ邮箱的自动回复邮件。邮件已收到，谢谢!

**Comment by enjoybo** [comment]
@karta2155802 Thanks for sharing your code. Perhaps I did not articulate my problem clearly. I now can render the mesh correctly using Pytorch3D. I just don't know how to reproduce the render result using opengl. The render result by opengl looks different with Pytorch3d, even though I'm using the same transform matrix.

**Comment by karta2155802** [comment]
I didn't use opengl for rendering. Sorry, I can't help you.

**Comment by enjoybo** [comment]
@karta2155802 That's all right, thanks.

**Comment by lhiceu** [comment]
Hi @karta2155802 ! ran into a strange problem when I wanted the same pytorch3d rendering results as you. 
I show the difference plot between the render mask (depth) and the observation camera's mask (depth), as shown in fig1,2. The plot indicates that they are not a match. 

![Figure_4](https://github.com/facebookresearch/pytorch3d/assets/46743006/689449aa-c667-4488-b116-57868e6609a2)
![Figure_6](https://github.com/facebookresearch/pytorch3d/assets/46743006/0890f840-11ad-4594-94dc-6a0938801fa6)

I transform mesh verts into camera coordinate system with its 6D pose as well as observation depth map with camera intrix. Their 3D points are a match as shown in fig3 using open3d.
![2023-09-18 11-01-15 的屏幕截图](https://github.com/facebookresearch/pytorch3d/assets/46743006/0daafa09-37e0-43a1-9fcb-d1468c64a819)


This is my code:
`def render(cad, intrix, pose, img_height, img_width, box, device):

    verts, faces = load_ply(cad)
    trg_mesh = Meshes(verts=[verts.to(device)], faces=[faces.to(device)])

    R = np.linalg.inv(pose[:,:3])
    T = pose[:,3]
    R = R[np.newaxis,:,:]
    T = T[np.newaxis,:]
    R = torch.from_numpy(R).to(device)
    T = torch.from_numpy(T).to(device)

    camera = np.array([intrix[0,0], intrix[1,1], intrix[0,2], intrix[1,2]])
    camera = torch.from_numpy(camera).to(device)

    fu, fv, cu, cv =camera.clone().type(torch.float32).cpu().numpy()

    cameras = PerspectiveCameras(
        R=R, 
        T=T,
        focal_length= torch.tensor([fu, fv], dtype=torch.float32).reshape(1,2).to(device),
        principal_point= torch.tensor([cu, cv], dtype=torch.float32).reshape(1,2).to(device), 
        in_ndc=False, 
        image_size=torch.tensor([img_height,img_width], dtype=torch.float32).reshape(1,2).to(device), 
        device=device)
    
    blend_params = BlendParams(sigma=1e-4, gamma=1e-4)
    raster_settings = RasterizationSettings(
            image_size=[img_height, img_width],
            blur_radius=np.log(1. / 1e-4 - 1.) * blend_params.sigma, 
            faces_per_pixel=100, 
        )
    
    rasterizer = MeshRasterizer(
            cameras=cameras,
            raster_settings=raster_settings,
        )
    
    silhouette_renderer = MeshRenderer(
        rasterizer = rasterizer,
        shader=SoftSilhouetteShader(blend_params=blend_params),
    )


    rasterizer_output = rasterizer(meshes_world = trg_mesh)
    depth = rasterizer_output.zbuf[0, :, :, 0]
    depth = torch.flip(depth, [0,1])

    silhouette = silhouette_renderer(meshes_world = trg_mesh, image_size=((img_height, img_width),),)
    silhouette = silhouette[0,...,3]
    silhouette = torch.flip(silhouette, [0,1])

    return depth, silhouette`

**Comment by enjoybo** [comment]
这是来自QQ邮箱的自动回复邮件。邮件已收到，谢谢!

**Comment by karta2155802** [comment]
I'm not sure where the mistake is. Make sure that the coordinates(directions of +x,+y,+z) are correct.
